blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,866 @@
+        msg.cid().map_err(|err| Error::Other(err.to_string()))
+    }
+
+    /// This is a helper to push that will help to make sure that the message fits the parameters
+    /// to be pushed to the MessagePool
+    pub fn add(&mut self, msg: &SignedMessage) -> Result<(), Error> {
+        let size = msg
+            .marshal_cbor()
+            .map_err(|err| Error::Other(err.to_string()))?
+            .len();
+        if size > 32 * 1024 {
+            return Err(Error::MessageTooBig);
+        }
+        if msg
+            .value()
+            .gt(&ToBigUint::to_biguint(&2_000_000_000).unwrap())
+        {
+            return Err(Error::MessageValueTooHigh);
+        }
+
+        self.verify_msg_sig(msg)?;
+
+        let tmp = msg.clone();
+        self.add_tipset(tmp, &self.cur_tipset.clone())
+    }
+
+    /// Add a SignedMessage without doing any of the checks
+    pub fn add_skip_checks(&mut self, m: SignedMessage) -> Result<(), Error> {
+        self.add_helper(m)
+    }
+
+    /// Return the string representation of the message signature
+    pub(crate) fn sig_cache_key(&mut self, msg: &SignedMessage) -> Result<String, Error> {
+        match msg.signature().signature_type() {
+            SignatureType::Secp256k1 => Ok(msg.cid().unwrap().to_string()),
+            SignatureType::BLS => {
+                if msg.signature().bytes().len() < 90 {
+                    return Err(Error::BLSSigTooShort);
+                }
+                let mut beginning = String::new();
+                beginning += &msg
+                    .cid()
+                    .unwrap()
+                    .to_bytes()
+                    .into_iter()
+                    .map(char::from)
+                    .collect::<String>();
+                beginning += &msg
+                    .signature()
+                    .bytes()
+                    .iter()
+                    .map(|b| char::from(*b))
+                    .collect::<String>();
+                Ok(beginning)
+            }
+        }
+    }
+
+    /// Verify the message signature. first check if it has already been verified and put into
+    /// cache. If it has not, then manually verify it then put it into cache for future use
+    fn verify_msg_sig(&mut self, msg: &SignedMessage) -> Result<(), Error> {
+        let sck = self.sig_cache_key(msg)?;
+        let is_verif = self.sig_val_cache.get(&sck);
+        match is_verif {
+            Some(()) => Ok(()),
+            None => {
+                let umsg = Cbor::marshal_cbor(msg.message())
+                    .map_err(|err| Error::Other(err.to_string()))?;
+                let verif = msg.signature().verify(umsg.as_slice(), msg.from());
+                match verif {
+                    Ok(()) => {
+                        self.sig_val_cache.put(sck, ());
+                        Ok(())
+                    }
+                    Err(value) => Err(Error::Other(value)),
+                }
+            }
+        }
```

Comment:
```suggestion
        let sck = self.sig_cache_key(msg)?;
        let is_verif = self.sig_val_cache.get(&sck);
        if let Some(()) = self.sig_val_cache.get(&sck) {
            return Ok(());
        }
        let umsg = msg.message().marshal_cbor()?;
        let verif = msg.signature().verify(umsg.as_slice(), msg.from())?;
        self.sig_val_cache.put(sck, ());
        Ok(())
```
Easier to just do this instead of commenting on individual things. I didn't add the error mappings because I'm hoping the conversion gets added